### DISCUSSION TITLE
How to disable internal WiFi module to use Edimax EW-7811Un WLAN dongle instead?

### DISCUSSION THREAD

**Discussion by firstchanceongit** [rern/rAudio#156] [General]
How can I disable the internal WiFi module on Raspberry Pi 3 Model B or later (e.g. in case of a poor connection due to metal case) in order to use an external Edimax dongle instead? A solution that works right from the first bootup would be preferred, if possible.

**Comment by rern** [comment]
There's a command in startup.sh script:
`(( $( rfkill | grep wlan | wc -l ) > 1 )) && rmmod brcmfmac`
which should disable on-board Wi-Fi if more than 1 `wlan` detected.
Try `rfkill` to see a wireless device list.

**Reply by firstchanceongit** [reply]
When I start with a fresh install (only preconfigured wifi file put in root directory (static IP and WEP)) and with only the Edimax dongle inserted, I cannot establish a browser connection via raudio.local. If I then add the LAN cable, I can access raudio.local. Under Networks -> WiFi I can see the WiFi data as per the wifi file - data seems to be all correct. (Under System -> Wireless WiFi is turned on.) Manual connect shows no effect.

rfkill returns with Edimax inserted:
1 wlan phy1 unblocked unblocked
and without:
0 wlan phy0 unblocked unblocked

Strangely, if I push the WiFi search button (magnifying glass), a few routers are shown (but not mine, even it is only a about a meter and an open door apart) without the Edimax dongle - but none at all if the Edimax dongle is inserted. WiFi is turned on on my router and other devices connect just fine to it.

EDIT:
I just checked - my router uses WPA+WPA2 security, not WEP. Could that be the reason? But how then to establish a connection? (Your WiFi profile generator does not support this.)

**Comment by R101G** [comment]
If you want to make sure the Edimax is wlan0, I would suggest this:
Add the line _dtoverlay=disable-wifi_ to /boot/config.txt otherwise the on-board wifi usually gets it.

In /boot/cmdline.txt add a space and _net.ifnames=0_ to the end of the line. It has to be on the same line, and stops the predictable naming, so that the first wifi device found will be called wlan0.

Also, for the Edimax, add a file to /etc/modprobe.d called blacklist-rtl8192cu.conf containing the line _blacklist rtl8192cu_. This is to stop an obsolete driver being used, which for some reason is still in the Arch kernel.

Do _systemctl enable netctl-auto@wlan0_ command to make sure it starts after a reboot.

If you still have problems, try ifconfig to see if the interface exists, and iwconfig to see if it is connected.

Also, check that the interface named in your /etc/netctl file is wlan0. If the system created it, it might not be.

EDIT: I have just checked, and it looks as if the latest kernel does not have the newer driver 8192cu, but does have the older rtl8192cu and rtl8xxxu drivers. This might mean something has been fixed, so try the Edimax without the blacklist file I suggested. If it connects, but does not work well, report back.

**Reply by firstchanceongit** [reply]
Thank you. But it seems that a basic point is not yet answered (or is at least unclear to me): Does the current image support WPA2 wifi connections at all?

**Comment by R101G** [comment]
Yes, that's the default if you don't specify WEP.

EDIT: Please check my edit to the earlier post!

**Comment by firstchanceongit** [comment]
Thanks to all. Did it as follows:

Image to SD card.
wifi file with static IP in root directory, config.txt and cmdline.txt adjusted as stated above.
Boot with Edimax only (no LAN connected) - -> works fine.
Power off.
Boot with Edimax removed, but LAN connected, then change to static IP via menu. -> works fine as well.

Not sure how later changes of the static IP for WLAN would work - just a new wifi file into the root directory prior to boot? Or via menu - but problem here that not the full static IP is shown, i.e. in the wifi file I used the following format for the IP=('xxx.xsx.xxx.xxx/24'), in the menu it shows the IP (different to the IP in the LAN menu) INCLUDING the brackets and the apostrophes, but last two characters (second apostrophe and closing bracket) are not shown.

Adding the suggested file in modprobe.d results in an error message when tapping on "Network". Removing the file removes the error message and everything works fine again. (But this was tested before I read the edit in the respective post above - so, seems to confirm the edit, no need for blacklisting (at least in my setup).)

Clarification in the web interface that WPA is the standard, unless WEP is switched on, might be helpful for other users as well.

**Reply by R101G** [reply]
There has always been confusion over rtl8192cu drivers. If 8192cu is present, it is best to blacklist the two old drivers. One of them might work, but if you find there are a lot of errors in iwconfig, or the connection is unreliable, blacklist both of them.

I find the best way to alter the existing wifi settings is to edit the /etc/netctl/SSID file.

The rAudio wifi settings work well with the onboard wifi, but you are on your own if you want to use a dongle. Linux support for wifi is pretty patchy, so you are lucky if you don't actually have to install a driver.

**Reply by firstchanceongit** [reply]
Thank you, again. Will then probably follow that route when cloning the SD card for a second player and adjust IP manually.

**Reply by firstchanceongit** [reply]
I have to correct myself: If I reboot with LAN cable disconnected and Edimax re-inserted again, I cannot connect via WiFi (only vie ethernet, if LAN cable reconnected.

In "System", WiFi is still switched on. (Switching it off and on again and reboot brings no different result.)

**Reply by R101G** [reply]
This might be the dodgy driver playing up.
Unfortunately, my experience is that switching the system wifi setting off and on completely messes things up in ways I do not understand, so it might mean starting again. I don't use the Edimax dongle, but I will flash a spare SD card with the latest image tomorrow, and see how it works for me. In the meantime, you could use ifconfig, iwconfig, and lsmod to see what is going on.

**Comment by R101G** [comment]
This might be the driver playing up, but my experience is that switching the wifi setting off and on messes things up in a way I do not understand. It might require starting from scratch.

I have looked for a driver download which is easy to install, but as all the other Pi distros have the 8192cu driver already installed and the faulty drivers disabled by default, I cannot find one.

My previous image kernel which has the 8192cu driver is 5.4.83 - 1. The current image kernel is 5.10.11- 4, but I don't know how to find out exactly when the driver went missing. I think the easiest solution is to ask rern to make available an earlier image.

@rern 
Is that possible?

**Reply by firstchanceongit** [reply]
How yould I check which driver is being used? I made a fresh install on the SD card (incl. the cmdline.txt and config.txt adjustments as well as the wifi file in the root directory, startup with Edimax (and no LAN cable), Until here it works. But no idea which driver is being used now.

I also downloaded from the Edimax webpage the linux driver package for my Edimax EW-7811Un. But now I'm sitting on a gz.file and have no idea what to do with it and how to install it. (Can't really say that the wiki.archlinux.org/index.php/Network_configuration/ pages are self-explaining to newbies ...)

**Reply by R101G** [reply]
The command lsmod will show all the loaded driver modules, but the reason the 8192cu driver was created is that the others do not work reliably.

I do not know enough to create a build and install procedure so, unless someone else does, using an older kernel until the Arch Linux gods correct the error seems to be the best solution. You can check the driver is present with _modprobe 8192cu_. If it returns an error, it's not there. If you want to try a different dongle instead, I know that an rtl8822bu driver can be installed, because that is what I use.

**Reply by firstchanceongit** [reply]
Thanks - I just came across the following instruction from the Edimax support team, and apparently, the driver package from the Edimax webpage will not help at all:

https://edimax.freshdesk.com/support/solutions/articles/14000035492-how-to-resolve-ew-7811un-built-in-driver-issues-in-linux-kernel-v3-10-or-higher

Unfortunately, the instruction works only for Debian based systems (and the alternative driver is deprecated (but still seems to work for some). I've updated the instruction to work for Archlinux instead for Debian:

1. Plugin the EW-7811Un to your system's USB port (2.0 preferred)
2. Make sure your system has temporary Internet connection
3. Open a Terminal program
4. Make sure your system is up-to-date
    	pacman -Syu
    [ Note:  You may need to restart your system if kernel update has been installed. ]
5.  Install the necessary files to compile a device driver
	sudo pacman -S base-devel
		when being asked, confirm: all
	sudo pacman -S dkms
	sudo pacman -S git
	sudo pacman -Sy linux-headers
		when being asked, confirm: confirm the respective number (e.g. "12" for a Raspberry Pi 3 Model B)
6.  Restart your system
    	sudo reboot
7.  After finished booting up, open a Terminal program again
8.  Clone the github repository
	git clone https://github.com/pvaret/rtl8192cu-fixes.git
    [ Note:  Right after "rt", it is a letter "l" (like Larry), not a number "1" (like 1980.)  Typing it wrong will cause GitHub to ask for Username and Password.  Please note that accordingly. ]
9.  Setup as DKMS module
	sudo dkms add ./rtl8192cu-fixes
10. Build and install the rtl8192cu driver
	sudo dkms install 8192cu/1.11
11. Refresh modules list
	sudo depmod -a
12. Blacklisted the native kernel
	sudo cp ./rtl8192cu-fixes/blacklist-native-rtl8192.conf /etc/modprobe.d/
13. Disable power management
	sudo cp ./rtl8192cu-fixes/8192cu-disable-power-management.conf /etc/modprobe.d/
14. Restart your system to use the new driver
	sudo reboot

NOTE: I got three warnings in Step 4 - /boot/config.txt, /boot/cmdline.txt and /etc/php/php.ini were saved on the SD card with the filename addition ".pacnew". So, in an additional Step 4a, I manually changed the original files config.txt, cmdline.txt and php.ini by appending "_OLD" to their filename and removed the ".pacnew" from the new versions in order to use them. Whether that was right or wrong - future will show ... ;-) (An alternative to Steps 4/4a could be to build the entire rAudio-Image from scratch based on the latest release from Arch Linux Arm via rern's rOS DIY rAudio (https://github.com/rern/rOS) and then to perform Steps 1-3, then 5-14.)

Whether the above will eventually lead to a stable WiFi connection - even after switching back and forth between LAN and WLAN - I cannot (yet) say. I guess this will also depend on how the interface of rAudio interacts with all of this - and possibly only rern will be able to answer this (or fix it, as may be necessary ...).

**Reply by R101G** [reply]
There is a dkms package if you can get it to work. Try _pacman -S dkms-8192cu_

**Reply by firstchanceongit** [reply]
Thank you. Unfortunately, this reports an error (at the end of the installation process):

:: Running post-transaction hooks...
(1/3) Arming ConditionNeedsUpdate...
**_(2/3) Install DKMS modules
==> Unable to install module 8192cu/v4.0.2_9000 for kernel *: Missing kernel headers._**
(3/3) Updating the info directory file...

So, I guess this will not work either :-(

**Reply by R101G** [reply]
I think they have fixed it. Try updating with _pacman -Syu_ and do _modprobe 8192cu_ Edit - No, They haven't.

If not, the instructions you found do work.

**Reply by firstchanceongit** [reply]
Updating as suggested didn't work in this regard:

**_(12/20) Install DKMS modules
==> Unable to install module 8192cu/v4.0.2_9000 for kernel 5.10.13-3-ARCH: Missing kernel headers._**

(There were also three warnings as to /boot/config.txt, /boot/cmdline.txt and /etc/php.php.ini being overwritten - but I guess this can be sorted out.)

modprobe 8192cu
results in
**_modprobe: FATAL: Module 8192cu not found in directory /lib/modules/5.4.83-3-ARCH_**

So, I'll switch back to an image from before the updates and try hw far I can get with the Edimax instruction.

**Reply by R101G** [reply]
One of the things I tried did install the driver, so I will try and figure out what worked.

**Comment by R101G** [comment]
I tried this from a new install, and it installs the driver and blacklists the old ones. You will still need to alter the boot files and create a wifi file to connect.

pacman -Syu

pacman -S dkms    
pacman -S git
pacman -S linux-headers           (Enter 12)
pacman -S bc

reboot

git clone https://github.com/pvaret/rtl8192cu-fixes.git

dkms add ./rtl8192cu-fixes

dkms install 8192cu/1.11

depmod -a

cp ./rtl8192cu-fixes/blacklist-native-rtl8192.conf /etc/modprobe.d/

cp ./rtl8192cu-fixes/8192cu-disable-power-management.conf /etc/modprobe.d/

reboot

**Reply by firstchanceongit** [reply]
Great minds think alike ;-)
I just did basically the same - see my edit above which seems to have crossed your efforts, but thanks for your time and support so far! :-)

The only difference in your instruction is the "pacman -S bc". Do I need that?

**Reply by R101G** [reply]
Probably not!

**Comment by firstchanceongit** [comment]
Hm. Something seems to be still missing:

I put a "wifi" file in the boot directory (for the first time in this installation). Bootup with Edimax and without LAN cable - no connection. When I connect the LAN I cann access rAudio. WiFi is switched on under "System".

ifconfig shows no wlan

dmesg | grep firmware
shows:
[    0.433352] raspberrypi-firmware soc:firmware: Attached to firmware from 2021-01-18 17:44, variant start
[    0.443357] raspberrypi-firmware soc:firmware: Firmware hash is e5e540537282a28aa8ec29d9662d65e06d0d2e24
[   13.479392] rtl8192cu: Loading firmware rtlwifi/rtl8192cufw_TMSC.bin

The data from the "wifi" file is contained in a file named after the SSID in the /etc/netctl directory and also shown in the respective WiFi dialog box under "Network".

rfkill
shows:
ID TYPE      DEVICE      SOFT      HARD
 0 bluetooth hci0   unblocked unblocked
 2 wlan      phy1   unblocked unblocked

**Reply by R101G** [reply]
If you are sure the onboard wifi is disabled in config.txt, try systemctl enable netctl-auto@wlan0

EDIT: I have just done a quick test, and the rfkill result you get implies that the onboard wifi is not disabled in config.txt

**Reply by firstchanceongit** [reply]
Ah, sure - stupid me! I forgot to (again) ajdust config.txt accordingly (only did so with regard to cmdline.txt) following the general system update above ... - shame on me! I had also tried the systemctl ... but without the change in config.txt. 

Now it seems to work (and it appears that the systemctl ... is not required).

I did a couple of reboots with Edimax only, LAN only, again Edimax only, LAN and Edimax at the same time, again Edimax only ... and all appears to be reliable. (At least for the time being ...).

I'm eternally grateful for your help ...
